File: apps/ios/Packages/AMUXCore/Sources/AMUXCore/CloudAPI/ServerEndpoint.swift
import Foundation

/// Pre-auth "choose your server" support. Self-host deployments point the
/// app at their own Cloud API from the login screen; everything else the
/// app needs (MQTT broker address, feature config) is discovered from that
/// base URL via `/v1/config/bootstrap`, so the Cloud API address is the
/// only thing a user ever has to type.
public enum ServerEndpoint {
    /// Turns user input into a canonical Cloud API base URL: trims
    /// whitespace, defaults the scheme to https, drops trailing slashes,
    /// and requires a host. Returns nil for input that cannot name a
    /// server — a bad value must never be persisted, or it wedges the app
    /// at launch.
    public static func normalize(_ raw: String) -> URL? {
        var text = raw.trimmingCharacters(in: .whitespacesAndNewlines)
        guard !text.isEmpty else { return nil }
        if !text.contains("://") {
            text = "https://" + text
        }
        while text.hasSuffix("/") { text.removeLast() }
        guard let components = URLComponents(string: text),
              let scheme = components.scheme?.lowercased(),
              scheme == "https" || scheme == "http",
              let host = components.host, !host.isEmpty
        else { return nil }
        return components.url
    }

    public enum ProbeOutcome: Equatable, Sendable {
        case reachable
        case badStatus(Int)
        case unreachable(String)
    }

    /// Asks the candidate server for its public config. A 200 from
    /// `GET /v1/config/public` is proof enough that a TeamClu Cloud API
    /// lives at this base URL; no auth is required.
    public static func probe(_ baseURL: URL, session: URLSession = .shared) async -> ProbeOutcome {
        var request = URLRequest(url: baseURL.appendingPathComponent("v1/config/public"))
        request.httpMethod = "GET"
        request.timeoutInterval = 6
        do {
            let (_, response) = try await session.data(for: request)
            guard let http = response as? HTTPURLResponse else {
                return .unreachable("no HTTP response")
            }
            return http.statusCode == 200 ? .reachable : .badStatus(http.statusCode)
        } catch {
            return .unreachable(error.localizedDescription)
        }
    }
}

public extension CloudAPIConfigurationStore {
    /// The server the app talks to when no override is stored.
    static var bundledCloudAPIURL: String? {
        let value = SharedDefaults.services.cloudApiUrl?.trimmingCharacters(in: .whitespacesAndNewlines)
        return value?.isEmpty == false ? value : nil
    }

    static func hasCloudAPIURLOverride(in defaults: UserDefaults = .standard) -> Bool {
        let raw = defaults.string(forKey: cloudAPIURLKey)?.trimmingCharacters(in: .whitespacesAndNewlines)
        return raw?.isEmpty == false
    }

    /// Persists (or clears, when nil) the user-chosen Cloud API base URL.
    /// Callers pass a URL from `ServerEndpoint.normalize` — raw user text
    /// is not accepted here.
    static func setCloudAPIURLOverride(_ url: URL?, in defaults: UserDefaults = .standard) {
        if let url {
            defaults.set(url.absoluteString, forKey: cloudAPIURLKey)
        } else {
            defaults.removeObject(forKey: cloudAPIURLKey)
        }
    }
}

File: apps/expo/src/lib/cloud-api/server-endpoint.ts
/**
 * Pre-auth "choose your server" support. Mirrors iOS `ServerEndpoint` and the
 * desktop Custom server step: self-host users point the app at their own Cloud
 * API; MQTT and the rest are discovered from that base URL.
 */

export type ProbeOutcome =
  | { kind: "reachable" }
  | { kind: "badStatus"; status: number }
  | { kind: "unreachable"; reason: string };

const PROBE_TIMEOUT_MS = 6000;

/**
 * Turns user input into a canonical Cloud API base URL: trims whitespace,
 * defaults the scheme to https, drops trailing slashes, and requires a host.
 * Returns null for input that cannot name a server — a bad value must never be
 * persisted, or it wedges the app at launch.
 */
export function normalizeServerEndpoint(raw: string): string | null {
  let text = raw.trim();
  if (!text) return null;
  if (!text.includes("://")) {
    text = `https://${text}`;
  }
  while (text.endsWith("/")) {
    text = text.slice(0, -1);
  }

  let parsed: URL;
  try {
    parsed = new URL(text);
  } catch {
    return null;
  }

  const scheme = parsed.protocol.replace(/:$/, "").toLowerCase();
  if (scheme !== "https" && scheme !== "http") return null;
  if (!parsed.hostname) return null;

  // Pasted browser URLs often carry a query or fragment. The client
  // string-concatenates paths onto this base, so strip both.
  parsed.search = "";
  parsed.hash = "";
  let path = parsed.pathname;
  while (path.endsWith("/") && path.length > 1) {
    path = path.slice(0, -1);
  }
  parsed.pathname = path === "/" ? "" : path;

  // URL#toString keeps an empty path as "" after we clear "/" — good.
  // Prefer no trailing slash on the origin-only form.
  return parsed.toString().replace(/\/$/, "") || null;
}

/**
 * Asks the candidate server for its public config. A 200 from
 * `GET /v1/config/public` is proof enough that a TeamClu Cloud API lives at
 * this base URL; no auth is required.
 */
export async function probeServerEndpoint(
  baseUrl: string,
  args?: { fetchImpl?: typeof fetch; timeoutMs?: number },
): Promise<ProbeOutcome> {
  const normalized = baseUrl.replace(/\/+$/, "");
  if (!normalized) return { kind: "unreachable", reason: "empty URL" };

  const fetchImpl = args?.fetchImpl ?? fetch;
  const controller = new AbortController();
  const timer = setTimeout(
    () => controller.abort(),
    args?.timeoutMs ?? PROBE_TIMEOUT_MS,
  );

  try {
    const response = await fetchImpl(`${normalized}/v1/config/public`, {
      method: "GET",
      signal: controller.signal,
    });
    if (response.status === 200) return { kind: "reachable" };
    return { kind: "badStatus", status: response.status };
  } catch (error) {
    const reason =
      error instanceof Error && error.name === "AbortError"
        ? "timed out"
        : error instanceof Error
          ? error.message
          : "network error";
    return { kind: "unreachable", reason };
  } finally {
    clearTimeout(timer);
  }
}
